drivers/flash/flash_simulator.c

```diff
@@ -0,0 +1,273 @@
+		/* only pull bits to zero */
+		*(FLASH(offset + i)) &= *((u8_t *)data + i);
+	}
+
+	STATS_INCN(flash_sim_stats, bytes_written, len);
+
+#ifdef CONFIG_FLASH_SIMULATOR_SIMULATE_TIMING
+	/* wait before returning */
+	k_busy_wait(CONFIG_FLASH_SIMULATOR_MIN_WRITE_TIME_US);
+	STATS_INCN(flash_sim_stats, flash_write_time_us,
+		   CONFIG_FLASH_SIMULATOR_MIN_WRITE_TIME_US);
+#endif
+
+	return 0;
+}
+
+static void unit_erase(const u32_t unit)
+{
+	const off_t unit_addr = CONFIG_FLASH_SIMULATOR_BASE_OFFSET +
+				(unit * CONFIG_FLASH_SIMULATOR_ERASE_UNIT);
+
+	/* byte pattern to fill the flash with */
+	u8_t byte_pattern = 0xFF;
+
+	/* erase the memory unit by pulling all bits to one */
+	memset(FLASH(unit_addr), byte_pattern,
+	       CONFIG_FLASH_SIMULATOR_ERASE_UNIT);
+}
+
+static int flash_sim_erase(struct device *dev, const off_t offset,
+			   const size_t len)
+{
+	ARG_UNUSED(dev);
+
+	if (!flash_range_is_valid(dev, offset, len)) {
+		return -EINVAL;
+	}
+
+#ifdef CONFIG_FLASH_SIMULATOR_ERASE_PROTECT
+	if (flash_wp_is_set()) {
+		return -EACCES;
+	}
+#endif
+	/* erase operation must be aligned to the erase unit boundary */
+	if ((offset % CONFIG_FLASH_SIMULATOR_ERASE_UNIT) ||
+	    (len % CONFIG_FLASH_SIMULATOR_ERASE_UNIT)) {
+		return -EINVAL;
+	}
+
+	STATS_INC(flash_sim_stats, flash_erase_calls);
+
+	/* the first unit to be erased */
+	u32_t unit_start = (offset - CONFIG_FLASH_SIMULATOR_BASE_OFFSET) /
+			   CONFIG_FLASH_SIMULATOR_ERASE_UNIT;
+
+	/* erase as many units as necessary and increase their erase counter */
+	for (u32_t i = 0; i < len / CONFIG_FLASH_SIMULATOR_ERASE_UNIT; i++) {
+		ERASE_CYCLES_INC(unit_start + i);
+		unit_erase(unit_start + i);
+	}
+
+#ifdef CONFIG_FLASH_SIMULATOR_SIMULATE_TIMING
+	/* wait before returning */
+	k_busy_wait(CONFIG_FLASH_SIMULATOR_MIN_ERASE_TIME_US);
+	STATS_INCN(flash_sim_stats, flash_erase_time_us,
+		   CONFIG_FLASH_SIMULATOR_MIN_ERASE_TIME_US);
+#endif
+
+	return 0;
+}
+
+#ifdef CONFIG_FLASH_PAGE_LAYOUT
+static const struct flash_pages_layout flash_sim_pages_layout = {
+	.pages_count = CONFIG_FLASH_SIMULATOR_FLASH_SIZE,
+	.pages_size = CONFIG_FLASH_SIMULATOR_ERASE_UNIT,
+};
+
+static void flash_sim_page_layout(struct device *dev,
+				  const struct flash_pages_layout **layout,
+				  size_t *layout_size)
+{
+	*layout = &flash_sim_pages_layout;
+	*layout_size = 1;
+}
+#endif
+
+static const struct flash_driver_api flash_sim_api = {
+	.read = flash_sim_read,
+	.write = flash_sim_write,
+	.erase = flash_sim_erase,
+	.write_protection = flash_wp_set,
+	.write_block_size = CONFIG_FLASH_SIMULATOR_PROG_UNIT,
+#ifdef CONFIG_FLASH_PAGE_LAYOUT
+	.page_layout = flash_sim_page_layout,
+#endif
+};
+
+static int flash_init(struct device *dev)
+{
+	STATS_INIT_AND_REG(flash_sim_stats, STATS_SIZE_32, "flash_sim_stats");
+	memset(mock_flash, 0xFF, ARRAY_SIZE(mock_flash));
+	rand_seed = sys_rand32_get();
+
+	return 0;
+}
+
+DEVICE_AND_API_INIT(flash_simulator, "FLASH_SIMULATOR", flash_init, NULL, NULL,
```

Comment:
Yes, this wasn't considered initially but could be added. I don't see much value in doing that though. This is a simulator and in practice it will be used to test libraries/subsystems, in which case one flash device is probably enough.